accumulator/accumulate_signals_workflow.go

```diff
@@ -0,0 +1,167 @@
+	GreetingsList      []AccumulateGreeting
+	UniqueGreetingKeys map[string]bool
+}
+
+// GetNextTimeout returns the maximum time for a workflow to wait for the next signal.
+// fromFirstSignalTimeout and signalToSignalTimeout can be adjusted to wait for the right amount of time
+// This resets with Continue As New
+func (a *AccumulateGreeting) GetNextTimeout(ctx workflow.Context, timeToExit bool, firstSignalTime time.Time) (time.Duration, error) {
+	if firstSignalTime.IsZero() {
+		firstSignalTime = workflow.Now(ctx)
+	}
+	total := workflow.Now(ctx).Sub(firstSignalTime)
+	totalLeft := fromFirstSignalTimeout - total
+	if totalLeft <= 0 {
+		return 0, nil
+	}
+	if signalToSignalTimeout < totalLeft {
+		return signalToSignalTimeout, nil
+	}
+	return totalLeft, nil
+}
+
+// AccumulateSignalsWorkflow workflow definition
+func AccumulateSignalsWorkflow(ctx workflow.Context, greetings GreetingsInfo) (allGreetings string, err error) {
+	log := workflow.GetLogger(ctx)
+	var a AccumulateGreeting
+	if greetings.GreetingsList == nil {
+		greetings.GreetingsList = []AccumulateGreeting{}
+	}
+	if greetings.UniqueGreetingKeys == nil {
+		greetings.UniqueGreetingKeys = make(map[string]bool)
+	}
+	var unprocessedGreetings []AccumulateGreeting
+	var firstSignalTime time.Time
+	exitRequested := false
+
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	// Listen to signals in a different goroutine
+	workflow.Go(ctx, func(gCtx workflow.Context) {
+		for {
+			log.Debug("In workflow.go signals goroutine for{}")
+			selector := workflow.NewSelector(gCtx)
+			selector.AddReceive(workflow.GetSignalChannel(gCtx, "greeting"), func(c workflow.ReceiveChannel, more bool) {
+				c.Receive(gCtx, &a)
+				unprocessedGreetings = append(unprocessedGreetings, a)
+				log.Debug("Signal Received with text: " + a.GreetingText + ", more: " + strconv.FormatBool(more))
+
+				a = AccumulateGreeting{}
+			})
+			selector.AddReceive(workflow.GetSignalChannel(gCtx, "exit"), func(c workflow.ReceiveChannel, more bool) {
+				c.Receive(gCtx, nil)
+				exitRequested = true
+				log.Debug("Exit Signal Received")
+			})
+			log.Debug("Before select greeting  text: " + a.GreetingText)
+			selector.Select(gCtx)
+			log.Debug("After select, greeting  text: " + a.GreetingText)
+			if firstSignalTime.IsZero() {
+				firstSignalTime = workflow.Now(gCtx)
+			}
+		}
+	})
+
+	for !workflow.GetInfo(ctx).GetContinueAsNewSuggested() {
+		// Wait for Signal or timeout
+		timeout, err := a.GetNextTimeout(ctx, exitRequested, firstSignalTime)
+
+		if err != nil {
+			log.Warn("Error awaiting signal")
+			return "", err
+		}
+		if timeout <= 0 {
+			// do final processing? or just check for signal one more time
+			log.Debug("No time left for signals, checking one more time")
+		}
+
+		log.Debug("Awaiting for " + timeout.String())
+		gotSignalBeforeTimeout, _ := workflow.AwaitWithTimeout(ctx, timeout, func() bool {
```

Comment:
May be worth noting that you're adding two events for every task that has a signal here. You are adding a timer start and potentially a timer cancel. This is technically fine, but may be worth noting. Some users choose not to create a timer every time.

Reply:
I reworked this due to your feedback above about using a timer with a selector.